### PR TITLE
optimize: revise inlining costs

### PR DESCRIPTION
**Pull request by vtjnash**
Add a bonus for Intrinsics called with mostly constant arguments. We know that simple expressions like `x*1 + 0` will get optimized later by LLVM, and also likely fold into other expressions, so try to reflect that in the cost estimated earlier. Additionally rebalance some of the other costs to more accurately reflect what they take in assembly.

### CODE REVIEW

**Review by aviatesk — COMMENTED** [review]
Looks good to me.

**Comment by aviatesk** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/51599#issuecomment-1749815967) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/908a94e_vs_20a5fa7/report.md).

File: base/compiler/optimize.jl [hunk 1]
```diff
@@ -1031,11 +1031,36 @@ function statement_cost(ex::Expr, line::Int, src::Union{CodeInfo, IRCode}, sptyp
...
+                if minarg + 1 <= nargs <= maxarg + 1
+                    # With mostly constant arguments, all Intrinsics tend to become very cheap
+                    # and are likely to combine with the operations around them,
+                    # so reduce their cost by half.
+                    cost = T_IFUNC_COST[iidx]
+                    if cost == 0 || nargs < 3 ||
+                       (f === Intrinsics.cglobal || f === Intrinsics.llvmcall) # these hold malformed IR, so argextype will crash on them
+                        return cost
+                    end
+                    aty2 = widenconditional(argextype(ex.args[2], src, sptypes))
```
[the remark this answers is not in the document]
Comment (vtjnash) [reply base/compiler/optimize.jl:1046]: I was unsure if some external package might try to request inlining costs before we called this on everything, and I wanted to still give a consistent answer then. I think it is no-cost to be here?

**Comment by vtjnash** [comment]
The benchmark change here is whether we inline the call or dispatch the signature to call
```
broadcast(::typeof(+), ::Int64, ::NTuple{10, Float64}, ::Int64, ::Vararg{Any})
```

The  penalty for making that choice is higher than it should be, and poorly reflected in the way we compute the inlining cost heuristics, but that is not relevant to this PR.

**Comment by maleadt** [comment]
This change broke GPUCompiler's `always_inline` mode. MWE:

```julia
@inline @generated function sink(i::T) where {T}
    llvmcall_str = """%slot = alloca i64
                     store volatile i64 %0, i64* %slot
                     %value = load volatile i64, i64* %slot
                     ret i64 %value"""
    return :(Base.llvmcall($llvmcall_str, T, Tuple{T}, i))
end

@eval f(x) = $(foldl((e, _) -> :($sink($e) + $sink(x)), 1:100; init=:x))
g() = f(10)

include("newinterp.jl")
@newinterp AlwaysInline
Core.Compiler.OptimizationParams(::AlwaysInline) =
    Core.Compiler.OptimizationParams(; inline_cost_threshold=typemax(Int))
Base.code_ircode(g, Tuple{}; interp=AlwaysInline())
```

@vtjnash Are we doing something fishy here, or why was this affected by this PR?

**Comment by aviatesk** [comment]
Currently `OptimizationParams(; inline_cost_threshold=typemax(Int))` doesn't allow for "always inline", given we cap the inlining cost at `typemax(UInt16)` (, which is very confusing). We're gonna need #48257 to make this work.

**Pull request merged**